golem-worker-executor-base/src/durable_host/mod.rs

```diff
@@ -2476,47 +2572,207 @@ impl<Ctx: WorkerCtx> WasiHttpView for DurableWorkerCtxWasiHttpView<'_, Ctx> {
-                        .get_read_only_to(account_id, &key, &path)
+                        .get_read_only_to(account_id, &file.key, &path)
                         .await?;
-                    Ok::<_, GolemError>(Some((token, path)))
+                    Ok::<_, GolemError>((
+                        path,
+                        IFSWorkerFile::Ro {
+                            file,
+                            _token: token,
+                        },
+                    ))
                 }
                 ComponentFilePermissions::ReadWrite => {
                     debug!("Loading read-write file {}", path.display());
                     file_loader
-                        .get_read_write_to(account_id, &key, &path)
+                        .get_read_write_to(account_id, &file.key, &path)
                         .await?;
-                    Ok(None)
+                    Ok((path, IFSWorkerFile::Rw))
                 }
             }
         }
     });
+    Ok(HashMap::from_iter(try_join_all(futures).await?))
+}
 
-    let results = try_join_all(futures).await?;
+async fn update_filesystem(
+    current_state: &mut HashMap<PathBuf, IFSWorkerFile>,
+    file_loader: &Arc<FileLoader>,
+    account_id: &AccountId,
+    root: &Path,
+    files: &[InitialComponentFile],
+) -> Result<(), GolemError> {
+    enum UpdateFileSystemResult {
+        NoChanges,
+        Remove(PathBuf),
+        Replace { path: PathBuf, value: IFSWorkerFile },
+    }
+
+    let desired_paths: HashSet<PathBuf> = HashSet::from_iter(
+        files
+            .iter()
+            .map(|f| root.join(PathBuf::from(f.path.to_rel_string()))),
+    );
+
+    // We do this in two phases to make errors less likely. First delete all files that are no longer needed and then create
+    // new ones.
+    let futures_phase_1 = current_state.iter().map(|(path, file)| {
+        let path = path.clone();
+        let should_keep = desired_paths.contains(&path);
+        async move {
+            match file {
+                IFSWorkerFile::Ro { .. } if !should_keep => {
+                    tokio::fs::remove_dir(&path)
+                        .await
+                        .map_err(|e| GolemError::Runtime {
+                            details: format!(
+                                "Failed deleting file during update: {}, {e}",
+                                path.display()
+                            ),
+                        })?;
+                    Ok::<_, GolemError>(UpdateFileSystemResult::Remove(path))
+                }
+                _ => Ok(UpdateFileSystemResult::NoChanges),
+            }
+        }
+    });
+
+    let futures_phase_2 = files.iter().map(|file| {
+        let path = root.join(PathBuf::from(file.path.to_rel_string()));
+        let file = file.clone();
+        let permissions = file.permissions;
+        let file_loader = file_loader.clone();
+
+        let existing = current_state.get(&path);
+
+        async move {
```

Comment:
The rules here are a bit arbitrary, but I tried to be as predictable as possible